File: apps/desktop/src-tauri/src/models/agent.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageAttachment {
    pub data: String,
    pub media_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub file_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SlashCommandInfo {
    pub name: String,
    pub description: String,
    pub category: String,
    pub source: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentState {
    Idle,
    Thinking,
    Executing,
    AwaitingApproval,
    Completed,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClaudeLaunchOptions {
    pub provider: Option<String>,
    pub model: Option<String>,
    pub permission_mode: Option<String>,
    pub effort: Option<String>,
    pub agent: Option<String>,
    pub claude_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AgentEvent {
    #[serde(rename_all = "camelCase")]
    SystemInit {
        session_id: String,
        model: Option<String>,
        permission_mode: Option<String>,
        tools: Option<Vec<String>>,
    },
    #[serde(rename_all = "camelCase")]
    SessionMeta {
        provider: Option<String>,
        conversation_id: Option<String>,
        agent: Option<String>,
        effort: Option<String>,
        claude_path: Option<String>,
        slash_commands: Option<Vec<SlashCommandInfo>>,
    },
    #[serde(rename_all = "camelCase")]
    AssistantMessageStart {
        message_id: String,
        turn_id: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    AssistantMessageDelta {
        message_id: String,
        content_delta: String,
    },
    #[serde(rename_all = "camelCase")]
    AssistantMessageComplete {
        message_id: String,
        turn_id: Option<String>,
        content: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    ThinkingStart {
        message_id: String,
        turn_id: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    ReasoningDelta {
        content_delta: String,
        message_id: Option<String>,
        turn_id: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    ReasoningComplete {
        message_id: Option<String>,
        turn_id: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    ToolUseStart {
        tool_use_id: String,
        name: String,
        input: serde_json::Value,
        turn_id: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    ToolInputDelta {
        tool_use_id: String,
        input_delta: String,
    },
    #[serde(rename_all = "camelCase")]
    ToolProgress {
        tool_use_id: Option<String>,
        name: Option<String>,
        status: String,
    },
    #[serde(rename_all = "camelCase")]
    ToolResultDelta {
        tool_use_id: String,
        content_delta: String,
        is_error: Option<bool>,
    },
    #[serde(rename_all = "camelCase")]
    ToolResultComplete {
        tool_use_id: String,
        content: Option<String>,
        is_error: bool,
    },
    #[serde(rename_all = "camelCase")]
    ApprovalRequested {
        approval_id: String,
        tool_use_id: Option<String>,
        tool_name: String,
        input: Option<serde_json::Value>,
        detail: String,
    },
    #[serde(rename_all = "camelCase")]
    ApprovalResolved {
        approval_id: String,
        allow: bool,
    },
    #[serde(rename_all = "camelCase")]
    Status {
        state: AgentState,
        tool: Option<String>,
        tool_use_id: Option<String>,
        message_id: Option<String>,
        turn_id: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    Result {
        result_text: String,
        duration_ms: u64,
        total_cost_usd: f64,
        is_error: bool,
    },
    Raw {
        data: serde_json::Value,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AgentMode {
    Supervised,
    Assisted,
    FullAccess,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateAgentSessionRequest {
    pub cli_name: String,
    pub mode: AgentMode,
    pub working_directory: Option<String>,
    pub workspace_id: String,
    pub initial_prompt: String,
    pub claude: Option<ClaudeLaunchOptions>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentSessionInfo {
    pub id: String,
    pub cli_name: String,
    pub provider: Option<String>,
    pub display_name: String,
    pub mode: AgentMode,
    pub working_directory: Option<String>,
    pub workspace_id: String,
    pub conversation_id: Option<String>,
    pub is_alive: bool,
    pub created_at: String,
    pub model: Option<String>,
    pub permission_mode: Option<String>,
    pub agent: Option<String>,
    pub effort: Option<String>,
    pub claude_path: Option<String>,
    pub capabilities_loaded: Option<bool>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentEventPayload {
    pub session_id: String,
    pub event: AgentEvent,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentExitPayload {
    pub session_id: String,
    pub exit_code: Option<i32>,
}

File: packages/shared/src/types/agent.ts
export type AgentChatMode = "supervised" | "assisted" | "fullAccess";
export type AgentState = "idle" | "thinking" | "executing" | "awaiting_approval" | "completed" | "error";
export type AgentStreamState = "pending" | "streaming" | "completed" | "error";
export type AgentProvider = "claude" | "codex" | "aider" | "unknown";
export type ClaudePermissionMode = AgentChatMode;
export type ClaudeEffort = "low" | "medium" | "high";

export interface ClaudeLaunchOptions {
  provider?: "claude";
  model?: string;
  permissionMode?: ClaudePermissionMode;
  effort?: ClaudeEffort;
  agent?: string;
  claudePath?: string;
}

export interface CreateAgentSessionRequest {
  cliName: string;
  mode: AgentChatMode;
  workingDirectory?: string;
  workspaceId: string;
  initialPrompt: string;
  claude?: ClaudeLaunchOptions;
}

export interface AgentSessionInfo {
  id: string;
  cliName: string;
  provider?: AgentProvider;
  displayName: string;
  mode: AgentChatMode;
  workingDirectory?: string;
  workspaceId: string;
  conversationId?: string;
  isAlive: boolean;
  createdAt: string;
  model?: string;
  permissionMode?: string;
  agent?: string;
  effort?: ClaudeEffort;
  claudePath?: string;
  capabilitiesLoaded?: boolean;
}

export type ImageMediaType = "image/png" | "image/jpeg" | "image/gif" | "image/webp";

export interface ImageAttachment {
  /** base64-encoded image data (no data-uri prefix) */
  data: string;
  /** MIME type */
  mediaType: ImageMediaType;
  /** Optional display filename */
  fileName?: string;
}

export interface ContentBlock {
  type: "text" | "tool_use";
  text?: string;
  id?: string;
  name?: string;
  input?: Record<string, unknown>;
}

export type AgentEvent =
  | { type: "system_init"; sessionId: string; model?: string; permissionMode?: string; tools?: string[] }
  | {
      type: "session_meta";
      provider?: AgentProvider;
      conversationId?: string;
      agent?: string | null;
      effort?: ClaudeEffort | null;
      claudePath?: string | null;
      slashCommands?: SlashCommandInfo[];
    }
  | { type: "assistant_message_start"; messageId: string; turnId?: string }
  | { type: "assistant_message_delta"; messageId: string; contentDelta: string }
  | { type: "assistant_message_complete"; messageId: string; turnId?: string; content?: string }
  | { type: "thinking_start"; messageId: string; turnId?: string }
  | { type: "reasoning_delta"; contentDelta: string; messageId?: string; turnId?: string }
  | { type: "reasoning_complete"; messageId?: string; turnId?: string }
  | { type: "tool_use_start"; toolUseId: string; name: string; input: Record<string, unknown>; turnId?: string }
  | { type: "tool_input_delta"; toolUseId: string; inputDelta: string }
  | { type: "tool_progress"; toolUseId?: string; name?: string; status: string }
  | { type: "tool_result_delta"; toolUseId: string; contentDelta: string; isError?: boolean }
  | { type: "tool_result_complete"; toolUseId: string; content?: string; isError: boolean }
  | {
      type: "approval_requested";
      approvalId: string;
      toolUseId?: string;
      toolName: string;
      input?: Record<string, unknown>;
      detail: string;
    }
  | { type: "approval_resolved"; approvalId: string; allow: boolean }
  | { type: "status"; state: AgentState; tool?: string; toolUseId?: string; messageId?: string; turnId?: string }
  | { type: "result"; resultText: string; durationMs: number; totalCostUsd: number; isError: boolean }
  | { type: "raw"; data: unknown };

export interface AgentEventPayload {
  sessionId: string;
  event: AgentEvent;
}

export interface AgentExitPayload {
  sessionId: string;
  exitCode?: number;
}

export type SlashCommandCategory = "local" | "builtin" | "plugin" | "skill";

export interface SlashCommandInfo {
  /** Command name without leading "/" (e.g. "commit") */
  name: string;
  /** Human-readable description */
  description: string;
  /** Where the command comes from */
  category: SlashCommandCategory;
  /** Plugin name if category is "plugin" or "skill" */
  source?: string;
}
